doc/purple-remote.1.in
.TH purple-remote 1

.SH "NAME"
purple-remote v@VERSION@ \- Purple command-line remote control

.SH "SYNOPSIS"
.TP 5
\fBpurple-remote \fI[command]\fR \fI[parameters]\fR
.TP 5
\fBpurple-remote -h \fI[command]\fR

.SH "DESCRIPTION"
.PP
\fbpurple-remote\fR is a command-line interface for controlling an already-running instance of the Purple Instant Messaging client. It currently has only rudimentary support for launching AIM-style URIs, and closing the running instance, although further functionality is planned.

.SH "OPTIONS"
The following options are provided by \fBpurple-remote\fR:
.TP
.B \-h [\fIcommand\fB]
Display help for \fBpurple-remote\fR or a specific command (see below).

.SH "COMMANDS"
The following commands are implemented by \fBpurple-remote\fR:
.TP
.B quit
Closes the currently-running instance of \fBPurple\fR.
.TP
.B uri [URI]
Sends the given AIM-style URI to the currently running instance of \fBPurple\fR (see below).

.SH "URIs"
The following URIs are currently valid. These default to acting on the first AIM account you have in your list:
.TP
.B aim:goim?screenname=[SCREENNAME]
Begins a new conversation with the given screenname.
.TP
.B aim:goim?screenname=[SCREENNAME]&message=[MESSAGE]
Sends the given message to the screenname. Spaces in the message must be replaced with +.
.TP
.B aim:gochat?roomname=[CHATROOM]
Opens a chat window in the given room.
.TP
.B aim:addbuddy?screenname=[SCREENNAME]
Opens a dialog to add the specified buddy to your buddy list.

.SH "FILES"
.TP
.I /tmp/purple_[USERNAME].[INSTANCE]
The UNIX domain socket used to contact the running instance of \fBPurple\fR.

.SH "SEE ALSO"
.BR purple (1),
.BR \fIhttp://purple.sourceforge.net/\fR,
.BR \fIhttp://www.sourceforge.net/projects/purple/\fR

.SH "AUTHOR"
.PP
This manpage was written by Robert McQueen <\fIrobot101@debian.org\fR>
for the Debian GNU/Linux system (but may be used by others).



NEWS
-=[ Gaim ]=-  The Pimpin' Penguin IM Client That's Good For The Soul!

1.5.1 (4/27/2007):
	Luke: There really isn't much in this release, some backported
	patches, and some patches from RedHat.  The only real purpose
	to this is to make it legal to release security fixes to this
	branch of the code.

1.5.0 (8/11/2005):
	Mark: No super crazy major changes here.  Just the usual bug
	fixes and some pretty important security updates.

	Stu: Buy Sean's book. He obviously needs the money. Other than that
	we fixed some bugs with this release (I don't think we did much else).
	SoC students are doing lots of cool stuff which we'll hopefully be
	able to bring you in a future release.

	Nathan: I've continued my streak of doing nothing useful lately.
	However, I've now got a brand-spanking-new DSL line, so I might
	actually get to contribute soon.  One way or another I'd like
	to get HEAD into a more useable state in the next month or two.
	Also, like Stu said, buy Sean's book.  I've been reading and
	fixing it for the last umpteen months, so you had better enjoy
	it.

1.4.0 (7/7/2005):
	Mark: The last month or four we've promoted a bunch of the Gaim
	Crazy Patch Writers to developers, so there is now an even larger
	team of brilliant and amazingly sexy committers working around
	the globe for your instant messaging pleasure.  Also, we have
	what I believe to be our first contribution from a Summer of Code
	student in this release: Jonathan Clark enabled the sending of
	files to certain ICQ users.  Support is still a bit	rough, but
	he'll be working on it throughout the summer.

	Luke: Exciting times this summer as our Summer of Code interns
	start their projects.  This is mostly a bug fix release, with the
	ICQ file transfer that Mark mentioned and some buddy icon work
	being the only real new code.  Hence the extra week delay.  A
	big thanks to our translators who keep churning out updates even
	when we give them short notice as well.  Enjoy!

1.3.1 (6/9/2005):
	Sean: It's been a while since I've done one of these. Welcome to
	Gaim 1.3.1, "New Hyde Park." This is, again, another bugfix release
	but it comes with two keen announcements. First, I'd like to welcome
	Christopher O'Brien to the Gaim team. He has integrated his work on
	the Meanwhile project into Gaim, ensuring that Gaim 2.0.0 will include
	Sametime support. Also, we're participating in Google's Summer of
	Code, which you all should check out at 
	http://gaim.sourceforge.net/summerofcode/

1.3.0 (5/10/2005):
	Luke: This release fixes SILC for multiple accounts.  HOWEVER, you
	may lose your buddy list (for SILC only) upgrading.  This is
	rather unavoidable as the previous code did not keep track of
	which account each buddy belonged to.

	Stu: I'm glad we're finally getting this out, if fixes a number of bugs
	ranging from minor to not so minor. I'd also like to welcome Gary to
	the team - he's done a great job with Guifications, I'm sure he'll do
	good stuff here too.

	Gary: Well I got pretty much nothing done for this release.  Although I
	do have quite a few things in the works that will come to fruition when
	I find some more free time.

1.2.1 (4/3/2005):
	Luke:  Several important fixes this time around.  Big thanks to 
	Robert McQueen, Stu, Nathan, Ethan, and everyone else who has helped 
	with this effort.  Maybe next time we can get a decent chance to fix the 
	problems *before* they go public to have a normal release process.

	Nathan: Jabber got some updates this release (finally).  I've got
	more planned, but simply haven't had the time to anything about it.

1.2.0 (3/17/2005):
	Luke: Happy St. Patrick's Day all.  Sean scheduled a release for
	today, we'll see if we can pull it off. :-)  This release
	features somewhat more than just bug fixes, some improvements
	have been made to the conversation API which may affect plugin
	developers.  Yahoo users should also thank Tim and Bleeter for
	their efforts.

	Etan: I'm going to NEWS since I actually did something this time.
	Jabber will allow you to unsubscribe to someone's presence
	without logging out and in again, this does break seeing yourself
	on your buddy list, but hopefully we'll fix that again for the
	next release.

	Stu: Happy St. Guinness^wPatrick's day.

1.1.4 (2/24/2005):
	Sean: Another bug fix release. A big thanks to Rob McQueen, Ari
	Pollak, Don Seiler, and Warren Togami: some packagers who helped
	debug a nasty glib 2.6 problem. Also thanks to our own Stu who
	found and destroyed an MSN crash and an HTML parsing error

	Luke: Early release to handle the MSN and HTML crashes, as well
	as the glib crash.  Maybe that makes up for having had a 4 week
	period before the last release.  This one should be the nice
	stable release we meant 1.1.3 to be. Enjoy!

	Stu: I cleaned up the whitespace in this NEWS, it was bugging me.

1.1.3 (2/17/2005):
	Luke: Yet another bug fix release, many thanks to everyone who has
	helped to make gaim more stable!

	Stu: I fixed too many Yahoo HTTP proxy bugs, I should just go and
	write some core HTTP support that works better. A good all round
	bug fix release otherwise.

1.1.2 (1/20/2005):
	Luke: Another Bug fix release. This one featuring a fix to the HTTP
	Method for MSN users and other MSN fixes. A big thanks to Stu and
	Felipe Contreras for those. Stu also spent a long time in valgrind
	and so this brings you a Gaim release with fewer memory leaks. This
	release is on time primarily because I want to see the MSN fixes
	make it into Debian and Fedora before freezes. Expect further bug
	fix only releases to come at longer intervals, at Mark's request.

	Daniel: This is my first NEWS! (YaY) I didn't really do too much for
	this release.  There are a few bugfixes, mostly wingaim stuff. Oh, and
	I like the new XP System Tray icons.

	Sean: I'm down here today. I haven't really been paying too much
	attention to these boring bugfix releases, but I'd like to thank
	David and welcome him to the team. I'd also like to thank Steven

src/purple-remote.c
/*
 * purple-remote
 *
 * Gaim is the legal property of its developers, whose names are too numerous
 * to list here.  Please refer to the COPYRIGHT file distributed with this
 * source distribution.
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include "internal.h"

#include <getopt.h>

#include <purple-remote/remote.h>

/*To be implemented:
	     "       info                     Show information about connected accounts\n"
	     "       list                     Print buddy list\n"
	     "       ison                     Show presence state of your buddy\n"
	     "       convo                    Open a new conversation window\n"
	     "       send                     Send message\n"
	     "       add                      Add buddy to buddy list\n"
	     "       remove                   Remove buddy from list\n"
	     "       -m, --message=MESG       Message to send or show in conversation window\n"
	     "       -t, --to=SCREENNAME      Select a target for command\n"
	     "       -p, --protocol=PROTO     Specify protocol to use\n"
	     "       -f, --from=SCREENNAME    Specify screen name to use\n"
	     "       -q, --quiet              Send message without showing a conversation\n"
	     "                                window\n"
*/

static struct option longopts[] = {
	{"message", required_argument, NULL, 'm'},
	{"to",      required_argument, NULL, 't'},
	{"protocol",required_argument, NULL, 'p'},
	{"from",    required_argument, NULL, 'f'},
	{"quiet",   no_argument,       NULL, 'q'},
	{"help",    no_argument,       NULL, 'h'},
	{0,0,0,0}
};

struct remoteopts {
	char *command;
	char *uri;
	gboolean help, quiet;
	char *message, *to, *from;
	int protocol;
};
struct remoteopts opts;

/*
 * Prints a message to the terminal/shell/console.
 * We try to convert "text" from UTF-8 to the user's locale.
 * If that fails then UTF-8 is used as a fallback.
 *
 * If channel is 1, the message is printed to stdout.
 * if channel is 2, the message is printed to stderr.
 */ 
static void
message(char *text, int channel)
{
	char *text_conv = NULL,*text_output;
	GError *error = NULL;

	text_conv = g_locale_from_utf8(text, -1, NULL, NULL, &error);

	if (text_conv == NULL) {
		g_warning("%s\n", error->message);
		g_error_free(error);
	}

	text_output = (text_conv ? text_conv : text);

	switch (channel) {
		case 1:
			puts(text_output);
			break;
		case 2:
			fputs(text_output, stderr);
			break;
		default:
			break;
	}

	if (text_conv)
		g_free(text_conv);
}

static void
show_remote_usage(const char *name)
{
	char *text = NULL;

	text = g_strdup_printf(_("Usage: %s command [OPTIONS] [URI]\n\n"
		"    COMMANDS:\n"
		"       uri                      Handle AIM: URI\n"
		"       away                     Popup the away dialog with the default message\n"
		"       back                     Remove the away dialog\n"
		"       quit                     Close running copy of Gaim\n\n"
		"    OPTIONS:\n"
		"       -h, --help [command]     Show help for command\n"), name);

	message(text, 1);
	g_free(text);

	return;
}

int
get_options(int argc, char *argv[])
{
	int i;

	memset(&opts, 0, sizeof(opts));
	opts.protocol = -1;

	while ((i=getopt_long(argc, argv, "m:t:p:f:qh", longopts, NULL)) != -1) {
		switch (i) {
		case 'm':
			opts.message = optarg;
			break;
		case 't':
			opts.to = optarg;
			break;
		case 'p':
			/* Do stuff here. */
			break;
		case 'f':
			opts.from = optarg;
			break;
		case 'q':
			opts.quiet = TRUE;
			break;
		case 'h':
			opts.help = TRUE;
			break;
		}
	}
	
	/* We must have non getopt'ed argument-- the command */
	if (optind < argc)
		opts.command = g_strdup(argv[optind++]);
	else
		return 1;

	if (opts.help)
		return 0;

	/* And we can have another argument--the URI. */
	/* but only if we're using the uri command. */
	if (!strcmp(opts.command, "uri")) {
		if (argc-optind == 1)
			opts.uri = g_strdup(argv[optind++]);
		else
			return 1;
	} else if (optind == argc)
		return 0;
	else
		return 1;

	return 0;			
}

static int
send_generic_command(guchar type, guchar subtype) {
	int fd = 0;
	GaimRemotePacket *p = NULL;

	fd = gaim_remote_session_connect(0);
	if (fd < 0) {
		message(_("Pidgin not running (on session 0)\nIs the \"Remote Control\" plugin loaded?\n"), 2);
		return 1;
	}
	p = gaim_remote_packet_new(type, subtype);
	gaim_remote_session_send_packet(fd, p);
	close(fd);
	gaim_remote_packet_free(p);

	return 0;
}

static int
send_command_uri() {
	int fd = 0;
	GaimRemotePacket *p = NULL;

	fd = gaim_remote_session_connect(0);
	if (fd < 0) {
		message(_("Pidgin not running (on session 0)\nIs the \"Remote Control\" plugin loaded?\n"), 2);
		return 1;
	}
	p = gaim_remote_packet_new(CUI_TYPE_REMOTE, CUI_REMOTE_URI);
	gaim_remote_packet_append_string(p, opts.uri);
	gaim_remote_session_send_packet(fd, p);
	close(fd);
	gaim_remote_packet_free(p);

	return 0;
}

static void
show_longhelp( char *name, char *command)
{
	if (!strcmp(command, "uri")) {
		message(_("\n"
		       "Using AIM: URIs:\n"
		       "Sending an IM to a screen name:\n"
		       "	purple-remote uri 'aim:goim?screenname=Penguin&message=hello+world'\n"
		       "In this case, 'Penguin' is the screen name we wish to IM, and 'hello world'\n"
		       "is the message to be sent.  '+' must be used in place of spaces.\n"
		       "Please note the quoting used above - if you run this from a shell the '&'\n"
		       "needs to be escaped, or the command will stop at that point.\n"
		       "Also,the following will just open a conversation window to a screen name,\n"
		       "with no message:\n"
		       "	purple-remote uri 'aim:goim?screenname=Penguin'\n\n"
		       "Joining a chat:\n"
		       "	purple-remote uri 'aim:gochat?roomname=PenguinLounge'\n"
		       "...joins the 'PenguinLounge' chat room.\n\n"
		       "Adding a buddy to your buddy list:\n"
		       "	purple-remote uri 'aim:addbuddy?screenname=Penguin'\n"
			  "...prompts you to add 'Penguin' to your buddy list.\n"), 1);
	}

	else if (!strcmp(command, "quit")) {
		message(_("\nClose running copy of Pidgin\n"), 1);
	}

	else if (!strcmp(command, "away")) {
		message(_("\nMark all accounts as \"away\" with the default message.\n"), 1);
	}

	else if (!strcmp(command, "back")) {
		message(_("\nSet all accounts as not away.\n"), 1);
	}

	else {
		show_remote_usage(name);
	}
}

int main(int argc, char *argv[])
{
#ifdef ENABLE_NLS
	setlocale (LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	bind_textdomain_codeset(PACKAGE, "UTF-8");
	textdomain(PACKAGE);
#endif

	if (get_options(argc, argv)) {
		show_remote_usage(argv[0]);
		return 0;
	}

	if (!strcmp(opts.command, "uri")) {
		if (opts.help)
			show_longhelp(argv[0], "uri");
		else
			return send_command_uri();
	}

	else if (!strcmp(opts.command, "away")) {
		if (opts.help)
			show_longhelp(argv[0], "away");
		else
			return send_generic_command(CUI_TYPE_USER, CUI_USER_AWAY);
	}

	else if (!strcmp(opts.command, "back")) {
		if (opts.help)
			show_longhelp(argv[0], "back");
		else
			return send_generic_command(CUI_TYPE_USER, CUI_USER_BACK);
	}

	else if (!strcmp(opts.command, "quit")) {
		if (opts.help)
			show_longhelp(argv[0], "quit");
		else
			return send_generic_command(CUI_TYPE_META, CUI_META_QUIT);
	}

	else {
		show_remote_usage(argv[0]);
		return 1;
	}
	
	return 0;
}
